### PR TITLE
Migration: Legend

### PR DESCRIPTION
**Pull request by marvin19**
Improved Legend.

### CODE REVIEW

File: ts/Core/Legend/Legend.ts [hunk 1]
```diff
@@ -1758,6 +1767,42 @@ namespace Legend {
...
+    /**
+     * @private
+     */
+    export function compose(ChartClass: typeof Chart): void {
+        if (composedMembers.indexOf(ChartClass) === -1) {
+            composedMembers.push(ChartClass);
+            addEvent(ChartClass, 'rendering', function (): void {
+                const chart = this;
+                // Title
+                chart.setTitle();
```
Comment (TorsteinHonsi) [inline ts/Core/Legend/Legend.ts:1793]: This looks a bit suspicious - the chart title is not related to the legend 🤔

Comment (bre1470) [reply ts/Core/Legend/Legend.ts:1793]: I thought that it might be related to the box calculations.

Comment (TorsteinHonsi) [reply ts/Core/Legend/Legend.ts:1793]: Yes, but the way it is implemented now, there will not be a chart title without the Legend module?

File: ts/Core/Chart/Chart.ts [hunk 2]
```diff
@@ -2414,106 +2412,102 @@ class Chart {
...
-         * @type {Highcharts.Legend}
-         */
-        chart.legend = new Legend(chart, options.legend as any);
-
-        // Get stacks
-        if (chart.getStacks) {
-            chart.getStacks();
-        }
+            // Title
+            chart.setTitle();
```
Comment (bre1470) [inline ts/Core/Chart/Chart.ts:2418]: Here is is called without Legend, @TorsteinHonsi .

Comment (TorsteinHonsi) [reply ts/Core/Chart/Chart.ts:2418]: So it's called _twice_ when the Legend is imported? 🤔

Comment (bre1470) [reply ts/Core/Chart/Chart.ts:2418]: Yes, correct.

**Review by bre1470 — APPROVED** [review]
Thank you! Looks good. 👍

File: ts/Core/Chart/Chart.ts (hunk 2 above)
Comment (TorsteinHonsi) [reply ts/Core/Chart/Chart.ts:2418]: Okay, so the reason for this is that we wanted to wrap the _whole_ rendering logic inside one event handler, and this `setTitle` call was done before the legend, right?

If so, I think we should rather create an event call after setting the title but before `getMargins()`, where we could hook in the legend creation. 

In this PR, there's also an `afterRendering` event that is not used by any component.

Comment (bre1470) [reply ts/Core/Chart/Chart.ts:2418]: > Okay, so the reason for this is that we wanted to wrap the _whole_ rendering logic inside one event handler, and this `setTitle` call was done before the legend, right?

Yes, and it is needed. We tested with changed position and it started to fail.

> If so, I think we should rather create an event call after setting the title but before `getMargins()`, where we could hook in the legend creation.

Do you mean that we move `setTitle` in front of the default handler, or do we have no default handler at all?

> In this PR, there's also an `afterRendering` event that is not used by any component.

`render` and `afterRender` are hijacked by redraw and update. Therefor we had to go with `rendering` / `afterRendering`.

Comment (TorsteinHonsi) [reply ts/Core/Chart/Chart.ts:2418]: > do we have no default handler at all?

Yes, that's what I think. In this case we should have a detailed event that is called at just the right moment. Like we do for example with `beforeRender`

https://github.com/highcharts/highcharts/blob/a905c72df5599de95cf9a8fd040d59338dde530c/ts/Core/Chart/Chart.ts#L2707-L2711

**Review by TorsteinHonsi — APPROVED** [review]

**Pull request merged**